src/main/java/com/sourcegraph/permission/PermissionRouter.java

```diff
@@ -0,0 +1,159 @@
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+@Path("/permissions")
+@Component
+public class PermissionRouter {
+    @ComponentImport
+    private static RepositoryService repositoryService;
+    @ComponentImport
+    private static UserManager userManager;
+    @ComponentImport
+    private static UserService userService;
+    @ComponentImport
+    private static SecurityService securityService;
+
+    @Autowired
+    public PermissionRouter(RepositoryService repositoryService, UserManager userManager, UserService userService, SecurityService securityService) {
+        PermissionRouter.repositoryService = repositoryService;
+        PermissionRouter.userManager = userManager;
+        PermissionRouter.userService = userService;
+        PermissionRouter.securityService = securityService;
+    }
+
+    // The getUsersWithRepositoryPermission endpoint returns a roaring bitmap containing the IDs of all the users
+    // that is granted the specified permission to the specified repository.
+    //
+    // Ex. /permissions/users?repository=PROJECT_1/rep_1&permission=read
+    @GET
+    @Path("/users")
+    public Response getUsersWithRepositoryPermission(@Context HttpServletRequest request, @QueryParam("repository") String repo, @QueryParam("permission") String perm) throws IOException {
+        UserProfile profile = userManager.getRemoteUser(request);
+        if (profile == null || !userManager.isSystemAdmin(profile.getUserKey())) {
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+        }
+
+        Repository repository = getRepository(repo);
+        if (repository == null) {
+            return Response.status(Response.Status.NOT_FOUND).entity("No such repository: " + repository).build();
+        }
+
+        Permission permission = getRepositoryPermission(perm);
+        if (permission == null) {
+            return Response.status(Status.UNPROCESSABLE_ENTITY).build();
+        }
+
+        RoaringBitmap bitmap = new RoaringBitmap();
+
+        UserSearchRequest.Builder builder = new UserSearchRequest.Builder();
+        builder.repositoryPermission(repository, permission);
+        UserSearchRequest search = builder.build();
+
+        PageRequest pageRequest = new PageRequestImpl(0, 5000);
+        do {
+            Page<ApplicationUser> page = userService.search(search, pageRequest);
+            for (ApplicationUser user : page.getValues()) {
+                bitmap.add(user.getId());
+            }
+            pageRequest = page.getNextPageRequest();
+        } while (pageRequest != null);
+
+        byte[] backing = serialize(bitmap);
+        return backing != null ? Response.ok(backing).build() : Response.serverError().build();
```

Comment:
- can you include in the response why it failed for serverError?
- Will the response include a header for the size of the response?
- Can you include a content-type header in the response? If you are unsure you can use `application/octet-stream`. But I have another comment about some basic content negotiation so we can get json back for debugging purposes. In that case you will likely need to make up a mime type for roaring bitmaps.

Reply:
> Will the response include a header for the size of the response?
The response includes the header `content-length` (bytes). I also added a debug header for cardinality of the bitmap (`X-Debug-Count`).

>  In that case you will likely need to make up a mime type for roaring bitmaps.
Got it! I'll use `application/roaring-bitmap`.